### PR TITLE
Intelligent pipeline

### PR DESCRIPTION
**Pull request by kristianschneider**
This PR tries to make the pipeline quicker by checking what steps can be omitted based on the changed files

### CODE REVIEW

**Comment by BjarkeMeier** [comment]
I assume this should be a draft PR.
The title is not useful as a commit message.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -78,7 +123,8 @@ jobs:
       ARTIFACT_PATH: './source/databricks/dist'
 
   create_prerelease:
-    needs: [time_series_bundle_ingestor_ci]
+    needs: [time_series_bundle_ingestor_ci,wheel_ci, terraform_validate]
```
Comment (lasrinnil) [inline .github/workflows/ci.yml:126]: ,wheel_ci -> , wheel_ci

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -36,39 +75,45 @@ jobs:
...
+    needs: [check_change,terraform_validate]
+    if: needs.check_change.outputs.hasterraformfiles == 'true'
     uses: Energinet-DataHub/.github/.github/workflows/terraform-validate.yml@2.6.1
     with:
       TERRAFORM_WORKING_DIR_PATH: './build/databricks_streaming_cluster'
       TERRAFORM_VERSION: '1.1.6'
 
   time_series_bundle_ingestor_ci:
     needs: [ci_base, terraform_validate, dotnet_solution_time_series_ci]
+    if: ${{ always() && !cancelled() && (needs.dotnet_solution_time_series_ci.result == 'success' )}}
```
Comment (lasrinnil) [inline .github/workflows/ci.yml:108]: always() && !cancelled() <- What does this mean?

Comment (BjarkeMeier) [reply .github/workflows/ci.yml:108]: https://docs.github.com/en/actions/learn-github-actions/expressions#always
https://docs.github.com/en/actions/learn-github-actions/expressions#cancelled

Comment (lasrinnil) [reply .github/workflows/ci.yml:108]: Thanks @BjarkeMeier :D!

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -22,8 +22,47 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0  # OR "2" -> To retrieve the preceding commit.
+      - name: Get changed dotnet files
+        id: changed-files-dotnet
+        uses: tj-actions/changed-files@v17.3
+        with:
+          files: |
+            **/source/TimeSeries/**
```
Comment (BjarkeMeier) [inline .github/workflows/ci.yml:37]: There are files in the `/source` folder that relates to .NET.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -22,8 +22,47 @@ jobs:
...
+      - name: Get changed dotnet files
```
Comment (BjarkeMeier) [inline .github/workflows/ci.yml:38]:
```suggestion
      - name: Get changed Databricks files
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -22,8 +22,47 @@ jobs:
...
+        uses: tj-actions/changed-files@v17.3
+        with:
+          files: |
+            **/source/databricks/**            
+      - name: Get changed terraform files
+        id: changed-files-terraform
+        uses: tj-actions/changed-files@v17.3
+        with:
+          files: |
+            build//**
```
Comment (BjarkeMeier) [inline .github/workflows/ci.yml:49]: Perhaps?
```suggestion
            build/**
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -22,8 +22,47 @@ jobs:
...
+        uses: tj-actions/changed-files@v17.3
+        with:
+          files: |
+            **/source/TimeSeries/**
+      - name: Get changed dotnet files
+        id: changed-files-databricks
+        uses: tj-actions/changed-files@v17.3
+        with:
+          files: |
+            **/source/databricks/**
```
Comment (BjarkeMeier) [inline .github/workflows/ci.yml:43]: There are files in the repo root that affects Databricks code.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -78,7 +123,8 @@ jobs:
       ARTIFACT_PATH: './source/databricks/dist'
 
   create_prerelease:
-    needs: [time_series_bundle_ingestor_ci]
+    needs: [time_series_bundle_ingestor_ci,wheel_ci, terraform_validate]
+    if: ${{ always() && !cancelled() && (needs.time_series_bundle_ingestor_ci.result == 'success'  || needs.wheel_ci.result == 'success' || needs.terraform_validate_streaming_cluster.result == 'success' )}}
```
Comment (BjarkeMeier) [inline .github/workflows/ci.yml:127]: Why always?
Isn't `!cancelled()` default?
I would suggest removing this if-condition. Probably also applies to the other if-conditions.

**Comment by BjarkeMeier** [comment]
@kristianschneider, has this PR been abandoned?

File: .github/workflows/ci.yml (hunk 3 above)
Comment (kristianschneider) [reply .github/workflows/ci.yml:37]: Are we not just interested in Timeseries in this pipeline ?

File: .github/workflows/ci.yml (hunk 7 above)
Comment (lasrinnil) [inline .github/workflows/ci.yml:127]: Perhaps a summary of this If is in place to enlighten newcomers (and me) about its ups and downs.

**Comment by Renetnielsen** [comment]
@cla-bot check

**Comment by BjarkeMeier** [comment]
Closing as the work seems abandonned.

**Pull request closed without merging**